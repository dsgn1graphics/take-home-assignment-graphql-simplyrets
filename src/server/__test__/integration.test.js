const gql = require('graphql-tag');
const nock = require('nock');

const { constructTestServer } = require('./__utils');
const {
	mockListingsData,
	mockListingsDataTransformed,
	mockListingsByLocationSmall
} = require('../datasources/__mocks__/listings');
const { mockJWT } = require('../datasources/__mocks__/simplyretsApi.mock');

const SimplyretsAPI = require('../datasources/simplyrets');

const LISTINGS_LAYOUT_VIEW_QUERY = gql`
	query ListingsLayoutView_Query {
		listings(limit: 1) {
			listingId
			listPrice
			listDate
			geo {
				lat
				lng
			}
			property {
				acres
				type
				lotDescription
			}
			photos
		}
	}
`;

const GET_LISTING = gql`
	query getLising($mlsId: ID!) {
			listing(mlsId: $mlsId) {
			listingId
			listPrice
			listDate
			geo {
				lat
				lng
			}
			property {
				acres
				type
				lotDescription
			}
			photos    
		}
	}
`;

const GET_LISTINGS_BY_LOCATION = gql`
	query getListingsByLocation {
		listingsByLocation(city: "Houston", limit: 1) {
			address {
				state
				city
				postalCode
				streetName
				country
				streetNumberText
				streetNumber
				full
				unit
				crossStreet
			}
			mlsId
			listDate
			modified
			listPrice
			leaseType
		}
	}
`;

const mockContextRequest = {
	req: {
		headers: {
			authorization: mockJWT
		}
	}
}

describe('Queries', () => {
	describe('get listings data', () => {
		it('fetches a collection of listings;', async () => {
			const {
				server
			} = constructTestServer();
			const simplyretsAPI = server.config.dataSources().simplyretsAPI;

			simplyretsAPI.get = jest.fn(() => mockListingsDataTransformed);
			const res = await server.executeOperation(
				{ query: LISTINGS_LAYOUT_VIEW_QUERY },
				mockContextRequest
			);
			expect(res).toMatchSnapshot();
		});

		it('fetches a single listing when a mlsId is provided', async () => {
			const {
				server
			} = constructTestServer();
			const simplyretsAPI = server.config.dataSources().simplyretsAPI;

			simplyretsAPI.get = jest.fn(() => mockListingsDataTransformed[0]);
			const res = await server.executeOperation(
				{ query: LISTINGS_LAYOUT_VIEW_QUERY },
				mockContextRequest
			);
			expect(res).toMatchSnapshot();

		});

		it('fetches a collection of listings by location', async () => {
			const {
				server
			} = constructTestServer();
			const simplyretsAPI = server.config.dataSources().simplyretsAPI;

			simplyretsAPI.get = jest.fn(() => mockListingsByLocationSmall);
			const res = await server.executeOperation(
				{ query: GET_LISTINGS_BY_LOCATION },
				mockContextRequest
			);
			expect(res).toMatchSnapshot();
		})

		it('returns an error message when isAuth is false', async () => {
			const {
				server
			} = constructTestServer();
			const simplyretsAPI = server.config.dataSources().simplyretsAPI;

			simplyretsAPI.get = jest.fn(() => mockListingsDataTransformed);
			const res = await server.executeOperation(
				{ query: LISTINGS_LAYOUT_VIEW_QUERY },
				{req: {
					headers: {
						authorization: null
					}
				}}
			);
			expect(res).toMatchSnapshot();
		})
	});
})