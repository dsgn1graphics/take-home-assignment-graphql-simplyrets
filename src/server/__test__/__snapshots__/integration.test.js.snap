// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queries get listings data fetches a collection of listings by location 1`] = `
Object {
  "data": Object {
    "listingsByLocation": Array [
      Object {
        "address": Object {
          "city": "Houston",
          "country": "United States",
          "crossStreet": "Dunne Ave",
          "full": "74434 East Sweet Bottom Br #18393",
          "postalCode": "77096",
          "state": "Texas",
          "streetName": "East Sweet Bottom Br",
          "streetNumber": 74434,
          "streetNumberText": "74434",
          "unit": "18393",
        },
        "leaseType": "FullServ",
        "listDate": "2011-05-23 11:50:30",
        "listPrice": "$20,714,261.00",
        "mlsId": 1005192,
        "modified": "2015-11-14 9:57:26",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries get listings data fetches a collection of listings; 1`] = `
Object {
  "data": Object {
    "listings": Array [
      Object {
        "geo": Object {
          "lat": 29.689418,
          "lng": -95.474464,
        },
        "listDate": "2011-05-23 11:50:30",
        "listPrice": "$20,714,261.00",
        "listingId": "49699701",
        "photos": Array [
          "https://s3-us-west-2.amazonaws.com/cdn.simplyrets.com/properties/trial/home9.jpg",
          "https://s3-us-west-2.amazonaws.com/cdn.simplyrets.com/properties/trial/home-inside-9.jpg",
        ],
        "property": Object {
          "acres": null,
          "lotDescription": "Private Backyard",
          "type": "RES",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries get listings data fetches a single listing when a mlsId is provided 1`] = `
Object {
  "data": Object {
    "listings": Array [
      Object {
        "geo": Object {
          "lat": 29.689418,
          "lng": -95.474464,
        },
        "listDate": "2011-05-23 11:50:30",
        "listPrice": "$20,714,261.00",
        "listingId": "49699701",
        "photos": Array [
          "https://s3-us-west-2.amazonaws.com/cdn.simplyrets.com/properties/trial/home9.jpg",
          "https://s3-us-west-2.amazonaws.com/cdn.simplyrets.com/properties/trial/home-inside-9.jpg",
        ],
        "property": Object {
          "acres": null,
          "lotDescription": "Private Backyard",
          "type": "RES",
        },
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Queries get listings data returns an error message when isAuth is false 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Cannot read properties of null (reading 'id')],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
